packages/database/src/modelClasses/Entity.js

```diff
@@ -194,4 +236,112 @@ export class EntityModel extends DatabaseModel {
       shouldSetBounds ? [geojson, geojson, code] : [geojson, code],
     );
   }
+
+  async fetchAncestorDetailsByDescendantCode(...args) {
+    const cacheKey = `fetchAncestorDetailsByDescendantCode:${JSON.stringify(args)}`;
+    const [descendantCodes, hierarchyId, ancestorType] = args;
+    return this.runCachedFunction(cacheKey, async () => {
+      const ancestorDescendantRelations = await this.database.executeSqlInBatches(
+        descendantCodes,
+        batchOfDescendantCodes => [
+          `
+          SELECT descendant.code as descendant_code, ancestor.code as ancestor_code, ancestor.name as ancestor_name
+          FROM
+            ancestor_descendant_relation
+          JOIN
+            entity as ancestor on ancestor.id = ancestor_descendant_relation.ancestor_id
+          JOIN
+            entity as descendant ON descendant.id = ancestor_descendant_relation.descendant_id
+          WHERE
+            descendant.code IN (${batchOfDescendantCodes.map(() => '?').join(',')})
+          AND
+            ancestor_descendant_relation.entity_hierarchy_id = ?
+          AND
+            ancestor.type = ?
+          ORDER BY
+            generational_distance ASC
+        `,
+          [...batchOfDescendantCodes, hierarchyId, ancestorType],
+        ],
+      );
+      const ancestorDetailsByDescendantCode = {};
+      ancestorDescendantRelations.forEach(r => {
+        ancestorDetailsByDescendantCode[r.descendant_code] = {
+          code: r.ancestor_code,
+          name: r.ancestor_name,
+        };
+      });
+      return ancestorDetailsByDescendantCode;
+    });
+  }
+
+  async getRelationsOfEntity(...args) {
+    const cacheKey = `getRelationsOfEntity:${JSON.stringify(args)}`;
+    const [
+      entityId,
+      hierarchyId,
+      ancestorsOrDescendants,
+      entityTypeOfRelations,
+      immediateRelativesOnly,
+    ] = args;
+    const [joinTablesOn, filterByEntityId] =
+      ancestorsOrDescendants === ANCESTORS
+        ? ['ancestor_id', 'descendant_id']
+        : ['descendant_id', 'ancestor_id'];
+    return this.runCachedFunction(cacheKey, async () =>
+      this.findWithSql(
+        `
+        SELECT entity.*
+        FROM
+          entity
+        JOIN
+          ancestor_descendant_relation ON entity.id = ${joinTablesOn}
+        WHERE
+          ${filterByEntityId} = ?
+        AND
+          entity_hierarchy_id = ?
+        ${
+          entityTypeOfRelations
+            ? `
+        AND
+          entity.type = ?
+        `
+            : ''
+        }
+        ${
+          immediateRelativesOnly
+            ? `
+        AND
+          generational_distance = 1
+        `
+            : ''
+        }
```

Comment:
A comment I forgot to add - it may also be worth it extracting those statements to variables

Reply:
Fixed in part 3